hazelcast/src/main/java/com/hazelcast/instance/impl/executejar/ClientExecuteJarStrategy.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance.impl.executejar;
+
+import com.hazelcast.instance.impl.BootstrappedInstanceProxy;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.core.JobStatus;
+import com.hazelcast.jet.impl.util.ResettableConcurrentMemoizingSupplier;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static com.hazelcast.jet.core.JobStatus.NOT_RUNNING;
+import static com.hazelcast.jet.core.JobStatus.STARTING;
+import static com.hazelcast.jet.impl.util.Util.toLocalDateTime;
+import static com.hazelcast.jet.impl.util.Util.uncheckRun;
+
+public class ClientExecuteJarStrategy {
+
+    private static final ILogger LOGGER = Logger.getLogger(ClientExecuteJarStrategy.class.getName());
+    private static final int JOB_START_CHECK_INTERVAL_MILLIS = 1_000;
+    private static final EnumSet<JobStatus> STARTUP_STATUSES = EnumSet.of(NOT_RUNNING, STARTING);
+
+    /**
+     * This method needs to be fully synchronized because it shuts down the remembered object
```

Comment:
I don't understand why this mean that the method must be synchronized.

You can do `BootstrappedInstanceProxy instanceProxy = singleton.remembered();` before try block, then reference will be constant for this method - no need for any synchrionization